From 8a1d2abe5370dd2ab074ad3cc5bbb397498d76bd Mon Sep 17 00:00:00 2001
From: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
Date: Thu, 5 Dec 2013 15:52:42 +0000
Subject: [PATCH 062/351] ASoC: florida: select ADSP for compressed stream via
 codec DAI

When opening a compressed stream the code previously looped
through all ADSPs until it found one with a compressed-stream
firmware currently selected. This isn't really valid because
we can't randomly choose any ADSP, firmware and DAI combination.

Since firmwares are currently hardcoded to specific ADSPs,
this change selects the correct DSP based on the codec DAI.

Change-Id: I5cbbd569ba3cd8edd369ed116d36d0b4a401e0d5
Signed-off-by: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
Signed-off-by: Nikesh Oswal <Nikesh.Oswal@wolfsonmicro.com>
---
 sound/soc/codecs/florida.c |   20 +++++++++++++-------
 1 file changed, 13 insertions(+), 7 deletions(-)

diff --git a/sound/soc/codecs/florida.c b/sound/soc/codecs/florida.c
index 2c0bb1a..b340ead 100644
--- a/sound/soc/codecs/florida.c
+++ b/sound/soc/codecs/florida.c
@@ -1747,7 +1747,7 @@ static int florida_open(struct snd_compr_stream *stream)
 	struct snd_soc_pcm_runtime *rtd = stream->private_data;
 	struct florida_priv *florida = snd_soc_codec_get_drvdata(rtd->codec);
 	struct arizona *arizona = florida->core.arizona;
-	int i, ret = 0;
+	int n_adsp, ret = 0;
 
 	mutex_lock(&florida->compr_info.lock);
 
@@ -1756,20 +1756,26 @@ static int florida_open(struct snd_compr_stream *stream)
 		goto out;
 	}
 
-	for (i = 0; i < FLORIDA_NUM_ADSP; ++i) {
-		if (wm_adsp_compress_supported(&florida->core.adsp[i], stream)) {
-			florida->compr_info.adsp = &florida->core.adsp[i];
-			break;
-		}
+	if (strcmp(rtd->codec_dai->name, "florida-dsp-voicectrl") == 0) {
+		n_adsp = 2;
+	} else if (strcmp(rtd->codec_dai->name, "florida-dsp-trace") == 0) {
+		n_adsp = 0;
+	} else {
+		dev_err(arizona->dev,
+			"No suitable compressed stream for dai '%s'\n",
+			rtd->codec_dai->name);
+		ret = -EINVAL;
+		goto out;
 	}
 
-	if (!florida->compr_info.adsp) {
+	if (!wm_adsp_compress_supported(&florida->core.adsp[n_adsp], stream)) {
 		dev_err(arizona->dev,
 			"No suitable firmware for compressed stream\n");
 		ret = -EINVAL;
 		goto out;
 	}
 
+	florida->compr_info.adsp = &florida->core.adsp[n_adsp];
 	florida->compr_info.stream = stream;
 out:
 	mutex_unlock(&florida->compr_info.lock);
-- 
1.7.9.5

